File: nodeApi/gameService.js
const express = require('express');
const cors = require('cors');
const { possibleGameMoves, possibleGameResults } = require('./constants');

const app = express();
app.use(express.json());
app.use(cors()); // Enable CORS for all routes in the app

// Route to play the game
app.post('/play', (req, res) => {
    // Receive player's action from the UI service
    const { player } = req.body;
    const playerMove = possibleGameMoves[player];
    const computerAction = getComputerAction();
    const computerMove = possibleGameMoves[computerAction];
    // Determine the winner based on player's move and computer's move
    const gameResult = determineWinner(playerMove, computerMove);
    res.json({ result: gameResult, player, computer: computerAction });
});

// Route to get available choices
app.get('/choices', (_, res) => {
    res.json(possibleGameMoves);
});

// Route to get computer's choice
app.get('/choice', (_, res) => {
    const computerAction = getComputerAction();
    res.json({ id: computerAction, name: possibleGameMoves[computerAction] });
});

// Function to get computer's action
const getComputerAction = () => {
    // Simulating computer's action with random choice
    const randomIndex = Math.floor(Math.random() * (Object.keys(possibleGameMoves).length) + 1);
    return randomIndex;
}

// Function to determine the winner
const determineWinner = (playerMove, computerMove) => {
    // If player's move is the same as computer's move, it's a tie
    if (playerMove === computerMove) {
        return "It's a tie!";
    }
    // Determine winner based on possible game results
    return possibleGameResults[`${playerMove}-${computerMove}`] || 'Invalid moves!';
}

const startGameService = () => {
    const PORT = process.env.PORT || 5001;
    app.listen(PORT, () => {
        console.log(`Server is running on port ${PORT}`);
    });
}

module.exports = { startGameService };

File: pythonApi/gameService.py
from flask import Flask, request, jsonify
from flask_cors import CORS
import random


app = Flask(__name__)
CORS(app)  # Enable CORS for all routes in the app

possible_game_moves = ['Rock', 'Paper', 'Scissors', 'Lizard', 'Spock']

possible_game_results = {
    ('Rock', 'Paper'): 'You LOSE !',
    ('Rock', 'Scissors'): 'You WIN !',
    ('Rock', 'Lizard'): 'You WIN !',
    ('Rock', 'Spock'): 'You LOSE !',

    ('Paper', 'Rock'): 'You WIN !',
    ('Paper', 'Scissors'): 'You LOSE !',
    ('Paper', 'Lizard'): 'You LOSE !',
    ('Paper', 'Spock'): 'You WIN !',

    ('Scissors', 'Rock'): 'You LOSE !',
    ('Scissors', 'Paper'): 'You WIN !',
    ('Scissors', 'Lizard'): 'You WIN !',
    ('Scissors', 'Spock'): 'You LOSE !',

    ('Lizard', 'Rock'): 'You LOSE !',
    ('Lizard', 'Paper'): 'You WIN !',
    ('Lizard', 'Scissors'): 'You LOSE !',
    ('Lizard', 'Spock'): 'You WIN !',

    ('Spock', 'Rock'): 'You WIN !',
    ('Spock', 'Paper'): 'You LOSE !',
    ('Spock', 'Scissors'): 'You WIN !',
    ('Spock', 'Lizard'): 'You LOSE !',
}

@app.route('/play', methods=['POST'])
def play():
    # Receive player's move from the UI service
    player_move = request.json.get('move')
    
    computer_move = get_computer_move()

    print("computer_move element:", computer_move)
    print("player_move element:", player_move)
    
    # Determine the winner based on player's move and computer's move
    game_result = determine_winner(player_move, computer_move)
    
    return jsonify({'result': game_result})

def get_computer_move():
    # response = requests.post(THIRD_PARTY_URL) what should be used here, but for simplicity i'm just gonna use random
    random_index = random.randint(0, len(possible_game_moves) - 1)
    return possible_game_moves[random_index]

def determine_winner(player_move, computer_move):
    # If player's move is the same as computer's move, it's a tie
    if player_move == computer_move:
        return "It's a tie!"
    return possible_game_results.get((player_move, computer_move), 'Invalid moves!')

if __name__ == '__main__':
    app.run(debug=True, port=5001)